src/txn/txn_rollback_to_stable.c

```diff
@@ -208,36 +208,38 @@ __txn_abort_row_replace_with_hs_value(
         hs_upd = NULL;
     }
 
-    /*
-     * If we found a history value that satisfied the given timestamp, add it to the update list.
-     * Otherwise remove the key by adding a tombstone.
-     */
-    if (valid_update_found) {
-        WT_ERR(__wt_update_alloc(session, hs_value, &upd, &size, WT_UPDATE_STANDARD));
+    if (replace) {
+        /*
+         * If we found a history value that satisfied the given timestamp, add it to the update
+         * list. Otherwise remove the key by adding a tombstone.
+         */
+        if (valid_update_found) {
+            WT_ERR(__wt_update_alloc(session, hs_value, &upd, &size, WT_UPDATE_STANDARD));
 
-        /* Clear the transaction id when recovery is in progress. */
-        if (F_ISSET(S2C(session), WT_CONN_RECOVERING))
+            /* Clear the transaction id when recovery is in progress. */
+            if (F_ISSET(S2C(session), WT_CONN_RECOVERING))
+                upd->txnid = WT_TXN_NONE;
+            else
+                upd->txnid = hs_start.txnid;
+            upd->durable_ts = durable_ts;
+            upd->start_ts = hs_start.timestamp;
+
+            /*
+             * Set the flag to indicate that this update has been restored from history store for
+             * the rollback to stable operation.
+             */
+            F_SET(upd, WT_UPDATE_RESTORED_FOR_ROLLBACK);
+        } else {
+            WT_ERR(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
             upd->txnid = WT_TXN_NONE;
-        else
-            upd->txnid = hs_start.txnid;
-        upd->durable_ts = durable_ts;
-        upd->start_ts = hs_start.timestamp;
+            upd->durable_ts = WT_TS_NONE;
+            upd->start_ts = WT_TS_NONE;
+        }
```

Comment:
None of the txnid/ts fields need to be set. I guess it's not all that important, but on the other hand...

We already have `__upd_alloc_tombstone()` in `txn.i`, could we generalize it and use it more widely?

Maybe we need a refactor ticket, something like:
```
*** ../include/txn.i    Wed Feb 12 00:35:45 2020
--- -   Wed Feb 12 01:19:14 2020
*************** 
*** 756,776 ****
  } 
    
  /*
!  * __upd_alloc_tombstone --
!  *     Allocate a tombstone update at a given transaction id and timestamp.
   */    
  static inline int
! __upd_alloc_tombstone(
!   WT_SESSION_IMPL *session, WT_UPDATE **updp, uint64_t txnid, wt_timestamp_t start_ts)
  { 
!     size_t size;

!     WT_RET(__wt_update_alloc(session, NULL, updp, &size, WT_UPDATE_TOMBSTONE));
!     (*updp)->txnid = txnid;
!     /* FIXME: Reevaluate this as part of PM-1524. */
!     (*updp)->durable_ts = (*updp)->start_ts = start_ts;
!     F_SET(*updp, WT_UPDATE_RESTORED_FROM_DISK);
!     return (0);
  }   
      
  /*  
--- 756,779 ----
  } 
    
  /*
!  * __wt_upd_alloc_tombstone --
!  *     Allocate a tombstone update with default values.
   */
  static inline int
! __wt_upd_alloc_tombstone(WT_SESSION_IMPL *session, WT_UPDATE **updp)
  {
!     size_t notused;

!     /*
!      * The underlying allocation code clears memory, which is the equivalent of setting:
!      *
!      *    WT_UPDATE.txnid = WT_TXN_NONE;
!      *    WT_UPDATE.durable_ts = WT_TS_NONE;
!      *    WT_UPDATE.start_ts = WT_TS_NONE;
!      *    WT_UPDATE.prepare_state = WT_PREPARE_INIT;
!      *    WT_UPDATE.flags = 0;
!      */
!     return (__wt_update_alloc(session, NULL, updp, &notused, WT_UPDATE_TOMBSTONE));
  }

  /*
***************
*** 825,834 ****
          return (0);

      /* If there is no ondisk value, there can't be anything in the history store either. */
!     if (cbt->ref->page->dsk == NULL || cbt->slot == UINT32_MAX) {
!         WT_RET(__upd_alloc_tombstone(session, updp, WT_TXN_NONE, WT_TS_NONE));
!         return (0);
!     }

      buf.data = NULL;
      buf.size = 0;
--- 828,835 ----
          return (0);

      /* If there is no ondisk value, there can't be anything in the history store either. */
!     if (cbt->ref->page->dsk == NULL || cbt->slot == UINT32_MAX)
!         return (__wt_upd_alloc_tombstone(session, updp));

      buf.data = NULL;
      buf.size = 0;
***************
*** 855,861 ****
       */
      if (stop.txnid != WT_TXN_MAX && stop.timestamp != WT_TS_MAX && !WT_IS_HS(S2BT(session)) &&
        __wt_txn_visible(session, stop.txnid, stop.timestamp)) {
!         WT_RET(__upd_alloc_tombstone(session, updp, stop.txnid, stop.timestamp));
          return (0);
      }

--- 856,867 ----
       */
      if (stop.txnid != WT_TXN_MAX && stop.timestamp != WT_TS_MAX && !WT_IS_HS(S2BT(session)) &&
        __wt_txn_visible(session, stop.txnid, stop.timestamp)) {
!         WT_RET(__wt_upd_alloc_tombstone(session, updp));
!
!       (*updp)->txnid = stop.txnid;
!       /* FIXME: Reevaluate this as part of PM-1524. */
!       (*updp)->durable_ts = (*updp)->start_ts = stop.timestamp;
!       F_SET(*updp, WT_UPDATE_RESTORED_FROM_DISK);
          return (0);
      }
```
`

Reply:
Thanks. Created a refactor ticket to handle same.
https://jira.mongodb.org/browse/WT-5570